fix: set TORCH_NCCL_BLOCKING_WAIT

cdoern opened the pull request:
apparently according to: https://pytorch.org/docs/stable/distributed.html timeout is not honored unless TORCH_NCCL_BLOCKING_WAIT is set to 1

cdoern commented:
https://github.com/Mergifyio backport release-v0.10

JamesKunstle commented:
There were many lines changed in this PR because of indentation so it's kinda hard to know what the _actual_ change is. Could you point me in the right direction?

cdoern commented:
sure @JamesKunstle !

here is what I added:

1. 
```
    try:
        cached_nccl_blocking_wait: Union[str, None] = os.environ[
            "TORCH_NCCL_BLOCKING_WAIT"
        ]
    except KeyError:
        cached_nccl_blocking_wait = None
    finally:
        os.environ["TORCH_NCCL_BLOCKING_WAIT"] = "1"
   ```
   this sets the NCCL env var to 1 and caches the original value.

I then wrapped the training code in a try: and added 

```
    finally:
        if cached_nccl_blocking_wait is None:
            del os.environ["TORCH_NCCL_BLOCKING_WAIT"]
        else:
            os.environ["TORCH_NCCL_BLOCKING_WAIT"] = cached_nccl_blocking_wait
        torch.cuda.empty_cache()
        gc.collect()
```

which clears the CUDA cache, calls gc.collect, and sets the cached value of the env variable back.

JamesKunstle commented:
@cdoern The watchdog didn't honor the timeout anyway.

booxter reviewed (changes requested):
This should not be merged because:

- timeout patch it tries to fix is going to be reverted.
- disabling watchdog thread has negative implications.
- this is not fixing the CI issue in the first place since the root cause was disc space.

ktdreyer commented:
@JamesKunstle I think we can close this because we resolved the NCCL timeouts in https://github.com/instructlab/instructlab/issues/3321 . Is that right?
